### ISSUE TITLE
header.unset and use in access-log

### ISSUE THREAD

**Issue by brunohass** [h2o/h2o#1162]
Hello.

We use h2o as a proxy, and when we deliver the data to the client we need to remove some headers, but we use some of these headers in the access log. When we use header.unset, we are no longer able to use the header in the access log.

Example:

```
access-log:
  path: "| rotatelogs /path/to/access-log-file.%Y%m%d 86400"
  format: "%{my-custom-header}o"

header.unset: "my-custom-header"
```

We need to remove the header only on delivery to the client, and be able to capture the header in the access log.

Thank you!

**Comment by yannick** [comment]
i think currently this cannot easily be supported since the log happens on request disposal.

it's also related to https://github.com/h2o/h2o/issues/1133 
following that line of thought one possibility would be that we allow logging of arbitrary env variables and then implement a feature where you can export headers from backend calls.
imo a bit of a hack though.

**Comment by kazuho** [comment]
@yannick 
> following that line of thought one possibility would be that we allow logging of arbitrary env variables and then implement a feature where you can export headers from backend calls.

Agreed.

The other possibility will be to record the header set built by the handler, and make them referable using the logging specifiers.

Can somebody confirm whether if Apache supports logging via use of `<`; i.e. does `%<{my-custom-header}o` log the original request header? If such thing is supported by Apache, it might make more sense for us to support the same convention, by retaining the headers that were originally built.

**Comment by brunohass** [comment]
@kazuho 

> Can somebody confirm whether if Apache supports logging via use of <; i.e. does %<{my-custom-header}o log the original request header? If such thing is supported by Apache, it might make more sense for us to support the same convention, by retaining the headers that were originally built.

Really don't know, sorry.

In Haproxy one can define headers for use in logs by calling _capture_  in the config.

Example:

    declare capture response len 6
    http-response capture res.hdr(My-Custom-Header) id 0

    # Delete all my-custom headers
    rspidel ^My-Custom.*

    log-format %[capture.res.hdr(0)]

**Comment by rnsanchez** [comment]
Looking at the sources, what if `remove_header()` didn't literally remove the header, but instead moved the header being removed into secondary `h2o_headers_t` attributes in both structs `st_h2o_req_t` and `st_h2o_res_t` (the latter perhaps not needed)?

This way the headers wouldn't be immediately removed, but still available for components that might want to access them---such as logging them, but not transmitting them (they're already out of `->headers`).  Only those components that _want_ to make use of the full set of headers would be impacted (again, the access log in case a log is not found in the `headers` attribute), as well as the struct's destructor; everything else would keep unchanged/unaffected.

(I'm very new to h2o sources, so this proposal could be way out of line.)

**Comment by kazuho** [comment]
Let's get this fixed before we release 2.2.

**Comment by kazuho** [comment]
Implemented in the PR above. `%<{my-custom-header}o` would log the header emitted by the generator (e.g. fastcgi).
